File: native/android/orb/src/main/kotlin/com/earlbalai/orb/OrbIdentity.kt
/*
 * OrbIdentity.kt
 *
 * Seed -> identity. A straight port of sections 3 and 4 of src/orb.js, pinned by
 * ConformanceTest. See native/SPEC.md §2.
 */
package com.earlbalai.orb

import kotlin.math.exp
import kotlin.math.max
import kotlin.math.min
import kotlin.math.roundToInt

/** The four galaxy archetypes, in shader-index order. */
enum class OrbArchetype(val index: Int) {
  SPIRAL(0), NEBULA(1), CORE(2), DEEP(3);

  val displayName: String get() = name.lowercase()

  companion object {
    fun fromIndex(i: Int): OrbArchetype = values()[i]
    fun fromName(n: String): OrbArchetype? = values().firstOrNull { it.displayName == n.lowercase() }
  }
}

/**
 * The four agent states, in shader-index order. The order is the conversational loop:
 * idle -> listening -> thinking -> speaking.
 */
enum class OrbState(val index: Int) {
  IDLE(0), LISTENING(1), THINKING(2), SPEAKING(3);

  val displayName: String get() = name.lowercase()

  companion object {
    fun fromIndex(i: Int): OrbState = values()[i.coerceIn(0, 3)]
    fun fromName(n: String): OrbState? = values().firstOrNull { it.displayName == n.lowercase() }
  }
}

/** An RGB triple in 0..1. Alpha is never part of an orb colour. */
data class OrbColor(val r: Double, val g: Double, val b: Double) {
  init {
    require(r.isFinite() && g.isFinite() && b.isFinite()) { "[Orb] colour channel is not finite" }
  }

  /** `#rrggbb`, quantised the way the reference does (round-half-up to 8 bits). */
  val hex: String
    get() = "#%02x%02x%02x".format(q8(r), q8(g), q8(b))

  /** The colour after the 8-bit round trip the palette goes through. */
  internal val quantised: OrbColor
    get() = OrbColor(q8(r) / 255.0, q8(g) / 255.0, q8(b) / 255.0)

  companion object {
    val BLACK = OrbColor(0.0, 0.0, 0.0)
    val WHITE = OrbColor(1.0, 1.0, 1.0)

    private fun q8(v: Double): Int = (255 * OrbMath.clamp01(v)).roundToInt()

    /** `#rgb`, `#rrggbb`, `rgb`, `rrggbb`; null for anything else. */
    fun fromHex(hex: String): OrbColor? {
      var s = hex.trim().removePrefix("#")
      if (s.length != 3 && s.length != 6) return null
      if (!s.all { it.isDigit() || it.lowercaseChar() in 'a'..'f' }) return null
      if (s.length == 3) s = s.map { "$it$it" }.joinToString("")
      val v = s.toLong(16)
      return OrbColor(((v shr 16) and 0xff) / 255.0, ((v shr 8) and 0xff) / 255.0, (v and 0xff) / 255.0)
    }

    /** From an Android packed ARGB int. Alpha is discarded. */
    fun fromArgb(argb: Int): OrbColor =
      OrbColor(((argb shr 16) and 0xff) / 255.0, ((argb shr 8) and 0xff) / 255.0, (argb and 0xff) / 255.0)
  }
}

/** An anchor colour plus three luminance-compensated accents. */
class OrbPalette internal constructor(
  /** Degrees, 0 until 360. Null for a hand-built palette. */
  val hue: Double?,
  val anchor: OrbColor,
  val accents: List<OrbColor>,
) {
  /** A hand-built palette. */
  constructor(anchor: OrbColor, accents: List<OrbColor>) : this(null, anchor, accents) {
    require(accents.size == 3) { "[Orb] palette.accents must be exactly 3 colours" }
  }

  override fun equals(other: Any?): Boolean =
    other is OrbPalette && other.hue == hue && other.anchor == anchor && other.accents == accents
  override fun hashCode(): Int = (hue?.hashCode() ?: 0) * 31 + anchor.hashCode() * 7 + accents.hashCode()
  override fun toString(): String = "OrbPalette(hue=$hue, anchor=${anchor.hex}, accents=${accents.map { it.hex }})"

  companion object {
    /**
     * Build a palette around a hue in degrees. Yellows and greens are far brighter than
     * blues and violets at the same nominal lightness, so `k` nudges lightness up in
     * proportion to how dark the hue is.
     */
    fun ofHue(hue: Double): OrbPalette {
      val hin = if (hue.isFinite()) hue else 0.0
      val h = ((hin % 360.0) + 360.0) % 360.0
      val k = 0.2 * (1 - OrbIdentity.hueLuma(h))
      return OrbPalette(
        h,
        OrbIdentity.hsl(h, 0.85, 0.42 + k).quantised,
        listOf(
          OrbIdentity.hsl(h, 0.95, min(0.72, 0.6 + k)).quantised,
          OrbIdentity.hsl((h + 16) % 360.0, 0.8, min(0.82, 0.7 + k)).quantised,
          OrbIdentity.hsl((h + 34) % 360.0, 0.9, min(0.9, 0.8 + k)).quantised,
        ),
      )
    }
  }
}

/**
 * Everything a seed determines. Pure and stable: the same string is the same orb, on
 * every platform.
 */
class OrbIdentity(val seed: String) {
  /** Unsigned 32-bit FNV-1a hash, held in a Long. */
  val hash: Long = hashSeed(seed)
  val palette: OrbPalette = PALETTES[(hash % PALETTES.size).toInt()]
  val archetype: OrbArchetype = OrbArchetype.fromIndex(((hash ushr 16) % 4).toInt())
  /** Structural phase, seconds. Every per-orb variation in the shader comes off this. */
  val phase: Double = (hash % 6283) / 1000.0
  /** Starting spin angle, radians. */
  val spin: Double = phase * 3.7
  /** Starting clock, seconds, so two seeds are never in lockstep. */
  val timeOffset: Double = ((hash ushr 8) % 40009) / 100.0

  override fun toString(): String =
    "OrbIdentity(seed=$seed, hash=$hash, hue=${palette.hue}, archetype=${archetype.displayName}, phase=$phase)"

  companion object {
    /** The fourteen identity hues, unevenly spaced. */
    val HUES: List<Double> = listOf(13.0, 34.0, 125.0, 146.0, 166.0, 187.0, 208.0, 228.0, 249.0, 270.0, 290.0, 311.0, 332.0, 353.0)

    /** The fourteen built-in palettes, one per hue. */
    val PALETTES: List<OrbPalette> = HUES.map { OrbPalette.ofHue(it) }

    /** FNV-1a, 32-bit, over UTF-16 code units (what JS `charCodeAt` yields). Unsigned, as a Long. */
    fun hashSeed(seed: String): Long {
      var h = 0x811c9dc5.toInt()
      for (ch in seed) {
        h = h xor ch.code
        h *= 0x01000193
      }
      return h.toLong() and 0xffffffffL
    }

    /** HSL to RGB, 0..1, the CSS algorithm. */
    internal fun hsl(h: Double, s: Double, l: Double): OrbColor {
      val a = s * min(l, 1 - l)
      fun f(n: Double): Double {
        val k = (n + h / 30) % 12
        return l - a * max(-1.0, min(min(k - 3, 9 - k), 1.0))
      }
      return OrbColor(f(0.0), f(8.0), f(4.0))
    }

    /** Rec.601 luma of a fully saturated hue. */
    internal fun hueLuma(h: Double): Double {
      val c = hsl(h, 1.0, 0.5)
      return 0.299 * c.r + 0.587 * c.g + 0.114 * c.b
    }
  }
}

internal object OrbMath {
  fun clamp01(v: Double): Double = if (v < 0) 0.0 else if (v > 1) 1.0 else v
  fun clamp(v: Double, lo: Double, hi: Double): Double = if (v < lo) lo else if (v > hi) hi else v
  /** GLSL's smoothstep, exactly, so the CPU state basis matches `stateW()` in the shader. */
  fun smoothstep(e0: Double, e1: Double, x: Double): Double {
    val t = clamp((x - e0) / (e1 - e0), 0.0, 1.0)
    return t * t * (3 - 2 * t)
  }
  /** Frame-rate independent first-order lerp factor. */
  fun lerpRate(dt: Double, tau: Double): Double = if (dt > 0) 1 - exp(-dt / tau) else 0.0
}

File: native/apple/Sources/OrbKit/OrbIdentity.swift
//  OrbIdentity.swift
//  OrbKit
//
//  Seed -> identity. A straight port of section 3 and 4 of src/orb.js, pinned by
//  Tests/OrbKitTests/ConformanceTests.swift. See native/SPEC.md §2.

import Foundation

/// The four galaxy archetypes, in shader-index order.
public enum OrbArchetype: Int, CaseIterable, Sendable {
  case spiral = 0, nebula, core, deep

  public var name: String { ["spiral", "nebula", "core", "deep"][rawValue] }

  public init?(name: String) {
    guard let i = OrbArchetype.allCases.firstIndex(where: { $0.name == name }) else { return nil }
    self = OrbArchetype.allCases[i]
  }
}

/// The four agent states, in shader-index order. The order is the conversational loop:
/// idle -> listening -> thinking -> speaking; ordinary transitions are between neighbours.
public enum OrbState: Int, CaseIterable, Sendable {
  case idle = 0, listening, thinking, speaking

  public var name: String { ["idle", "listening", "thinking", "speaking"][rawValue] }

  public init?(name: String) {
    guard let i = OrbState.allCases.firstIndex(where: { $0.name == name }) else { return nil }
    self = OrbState.allCases[i]
  }
}

/// An RGB triple in 0...1. Alpha is never part of an orb colour: the orb composites
/// against whatever is behind it.
public struct OrbColor: Equatable, Sendable {
  public var r: Double
  public var g: Double
  public var b: Double

  public init(r: Double, g: Double, b: Double) {
    self.r = OrbMath.clamp01(r); self.g = OrbMath.clamp01(g); self.b = OrbMath.clamp01(b)
  }

  /// `#rgb`, `#rrggbb`, `rgb`, `rrggbb`. Returns nil for anything else.
  public init?(hex: String) {
    var s = hex.trimmingCharacters(in: .whitespaces)
    if s.hasPrefix("#") { s.removeFirst() }
    guard s.count == 3 || s.count == 6, s.allSatisfy({ $0.isHexDigit }) else { return nil }
    if s.count == 3 { s = s.map { "\($0)\($0)" }.joined() }
    let v = UInt32(s, radix: 16)!
    self.init(r: Double((v >> 16) & 0xff) / 255, g: Double((v >> 8) & 0xff) / 255, b: Double(v & 0xff) / 255)
  }

  /// `#rrggbb`, quantised the same way the reference does (round-half-up to 8 bits).
  public var hex: String {
    let q = { (v: Double) -> Int32 in Int32((255 * OrbMath.clamp01(v)).rounded(.toNearestOrAwayFromZero)) }
    return String(format: "#%02x%02x%02x", q(r), q(g), q(b))
  }

  /// The colour after the 8-bit round trip the palette goes through.
  var quantised: OrbColor {
    let q = { (v: Double) -> Double in (255 * OrbMath.clamp01(v)).rounded(.toNearestOrAwayFromZero) / 255 }
    return OrbColor(r: q(r), g: q(g), b: q(b))
  }

  public static let black = OrbColor(r: 0, g: 0, b: 0)
  public static let white = OrbColor(r: 1, g: 1, b: 1)
}

/// An anchor colour plus three luminance-compensated accents.
public struct OrbPalette: Equatable, Sendable {
  /// Degrees, 0..<360. Nil for a hand-built palette.
  public let hue: Double?
  public let anchor: OrbColor
  public let accents: [OrbColor]

  public init(anchor: OrbColor, accents: [OrbColor]) {
    precondition(accents.count == 3, "OrbPalette needs exactly 3 accents")
    self.hue = nil
    self.anchor = anchor
    self.accents = accents
  }

  init(hue: Double, anchor: OrbColor, accents: [OrbColor]) {
    self.hue = hue; self.anchor = anchor; self.accents = accents
  }

  /// Build a palette around a hue in degrees. Yellows and greens are far brighter than
  /// blues and violets at the same nominal lightness, so `k` nudges lightness up in
  /// proportion to how dark the hue is.
  public init(hue: Double) {
    let hin = hue.isFinite ? hue : 0
    let h = (hin.truncatingRemainder(dividingBy: 360) + 360).truncatingRemainder(dividingBy: 360)
    let k = 0.2 * (1 - OrbIdentity.hueLuma(h))
    self.init(
      hue: h,
      anchor: OrbIdentity.hsl(h, 0.85, 0.42 + k).quantised,
      accents: [
        OrbIdentity.hsl(h, 0.95, min(0.72, 0.6 + k)).quantised,
        OrbIdentity.hsl((h + 16).truncatingRemainder(dividingBy: 360), 0.8, min(0.82, 0.7 + k)).quantised,
        OrbIdentity.hsl((h + 34).truncatingRemainder(dividingBy: 360), 0.9, min(0.9, 0.8 + k)).quantised,
      ])
  }
}

/// Everything a seed determines. Pure and stable: the same string is the same orb, on
/// every platform.
public struct OrbIdentity: Equatable, Sendable {
  public let hash: UInt32
  public let seed: String
  public let palette: OrbPalette
  public let archetype: OrbArchetype
  /// Structural phase, seconds. Every per-orb variation in the shader comes off this.
  public let phase: Double
  /// Starting spin angle, radians.
  public let spin: Double
  /// Starting clock, seconds, so two seeds are never in lockstep.
  public let timeOffset: Double

  /// The fourteen identity hues, unevenly spaced.
  public static let hues: [Double] = [13, 34, 125, 146, 166, 187, 208, 228, 249, 270, 290, 311, 332, 353]

  /// The fourteen built-in palettes, one per hue.
  public static let palettes: [OrbPalette] = hues.map { OrbPalette(hue: $0) }

  /// FNV-1a, 32-bit, over UTF-16 code units (what JS `charCodeAt` yields).
  public static func hashSeed(_ seed: String) -> UInt32 {
    var h: UInt32 = 0x811c9dc5
    for unit in seed.utf16 {
      h ^= UInt32(unit)
      h = h &* 0x01000193
    }
    return h
  }

  public init(seed: String) {
    let h = OrbIdentity.hashSeed(seed)
    let phase = Double(h % 6283) / 1000
    self.hash = h
    self.seed = seed
    self.palette = OrbIdentity.palettes[Int(h % UInt32(OrbIdentity.palettes.count))]
    self.archetype = OrbArchetype(rawValue: Int((h >> 16) % 4))!
    self.phase = phase
    self.spin = phase * 3.7
    self.timeOffset = Double((h >> 8) % 40009) / 100
  }

  // MARK: colour math

  /// HSL to RGB, 0...1, the CSS algorithm.
  static func hsl(_ h: Double, _ s: Double, _ l: Double) -> OrbColor {
    let a = s * min(l, 1 - l)
    func f(_ n: Double) -> Double {
      let k = (n + h / 30).truncatingRemainder(dividingBy: 12)
      return l - a * max(-1, min(k - 3, 9 - k, 1))
    }
    return OrbColor(r: f(0), g: f(8), b: f(4))
  }

  /// Rec.601 luma of a fully saturated hue.
  static func hueLuma(_ h: Double) -> Double {
    let c = hsl(h, 1, 0.5)
    return 0.299 * c.r + 0.587 * c.g + 0.114 * c.b
  }
}

enum OrbMath {
  @inline(__always) static func clamp01(_ v: Double) -> Double { v < 0 ? 0 : (v > 1 ? 1 : v) }
  @inline(__always) static func clamp(_ v: Double, _ lo: Double, _ hi: Double) -> Double { v < lo ? lo : (v > hi ? hi : v) }
  /// GLSL's smoothstep, exactly, so the CPU state basis matches `stateW()` in the shader.
  @inline(__always) static func smoothstep(_ e0: Double, _ e1: Double, _ x: Double) -> Double {
    let t = clamp((x - e0) / (e1 - e0), 0, 1)
    return t * t * (3 - 2 * t)
  }
  /// Frame-rate independent first-order lerp factor.
  @inline(__always) static func lerpRate(_ dt: Double, _ tau: Double) -> Double { dt > 0 ? 1 - exp(-dt / tau) : 0 }
}
